test/test_response.py

```diff
@@ -24,13 +25,64 @@
 )
 from urllib3.response import (  # type: ignore[attr-defined]
     BaseHTTPResponse,
+    BytesQueueBuffer,
     HTTPResponse,
     brotli,
     zstd,
 )
 from urllib3.util.response import is_fp_closed
 from urllib3.util.retry import RequestHistory, Retry
 
+
+class TestBytesQueueBuffer:
+    def test_single_chunk(self) -> None:
+        buffer = BytesQueueBuffer()
+        assert len(buffer) == 0
+        with pytest.raises(ValueError, match="buffer is empty"):
+            assert buffer.get(10)
+
+        buffer.put(b"foo")
+        with pytest.raises(ValueError, match="n should be > 0"):
+            buffer.get(-1)
+
+        assert buffer.get(1) == b"f"
+        assert buffer.get(2) == b"oo"
+        with pytest.raises(ValueError, match="buffer is empty"):
+            assert buffer.get(10)
+
+    def test_read_too_much(self) -> None:
+        buffer = BytesQueueBuffer()
+        buffer.put(b"foo")
+        assert buffer.get(100) == b"foo"
+
+    def test_multiple_chunks(self) -> None:
+        buffer = BytesQueueBuffer()
+        buffer.put(b"foo")
+        buffer.put(b"bar")
+        buffer.put(b"baz")
+        assert len(buffer) == 9
+
+        assert buffer.get(1) == b"f"
+        assert len(buffer) == 8
+        assert buffer.get(4) == b"ooba"
+        assert len(buffer) == 4
+        assert buffer.get(4) == b"rbaz"
+        assert len(buffer) == 0
+
+    @pytest.mark.skipif(
+        sys.version_info < (3, 8), reason="pytest-memray requires Python 3.8+"
+    )
+    @pytest.mark.limit_memory("12.5 MB")  # assert that we're not doubling memory usage
+    def test_memory_usage(self) -> None:
+        # Allocate 10 1MiB chunks
+        buffer = BytesQueueBuffer()
+        for i in range(10):
+            # This allocates 2MiB, putting the max at around 12MiB. Not sure why.
+            buffer.put(bytes(2**20))
```

Comment:
Doesn't this allocate 1MB (2**20 = 1MB) making the max of 12MB make sense?

Reply:
memray tells me that it allocates 2MiB, and the whole test uses 12MiB. I would have expected 11MiB instead.